core/query/sql.ml

```diff
@@ -150,157 +150,196 @@ let order_by_clause n =
-      | Apply (f, args) -> buf_add f; buf_add "("; buf_mapstrcat buf args (fun x -> sb x) ","; buf_add ")"
-      | Empty q -> buf_add "not exists ("; string_of_query buf quote true q; buf_add ")"
-      | Length q -> buf_add "select count(*) from ("; string_of_query buf quote true q; buf_add ") as "; buf_add (fresh_dummy_var ())
-      | RowNumber [] -> buf_add "1"
-      | RowNumber ps ->
-        buf_add "row_number() over (order by "; buf_add (String.concat "," (List.map (string_of_projection quote one_table) ps)); buf_add ")"
+      | Apply ("not", [Empty q]) -> Format.fprintf ppf "exists (%a)" pr_q q
+      | Apply ("not", [v]) -> Format.fprintf ppf "not (%a)" pr_b v
+      | Apply (("negate" | "negatef"), [v]) -> Format.fprintf ppf "-(%a)" pr_b v
+      | Apply ("&&", [v; w]) -> Format.fprintf ppf "(%a) and (%a)" pr_b v pr_b w
+      | Apply ("||", [v; w]) -> Format.fprintf ppf "(%a) or (%a)" pr_b v pr_b w
+      | Apply ("==", [v; w]) -> Format.fprintf ppf "(%a) = (%a)" pr_b v pr_b w
+      | Apply ("<>", [v; w]) -> Format.fprintf ppf "(%a) <> (%a)" pr_b v pr_b w
+      | Apply ("<", [v; w]) -> Format.fprintf ppf "(%a) < (%a)" pr_b v pr_b w
+      | Apply (">", [v; w]) -> Format.fprintf ppf "(%a) > (%a)" pr_b v pr_b w
+      | Apply ("<=", [v; w]) -> Format.fprintf ppf "(%a) <= (%a)" pr_b v pr_b w
+      | Apply (">=", [v; w]) -> Format.fprintf ppf "(%a) >= (%a)" pr_b v pr_b w
+      | Apply ("RLIKE", [v; w]) -> Format.fprintf ppf "(%a) RLIKE (%a)" pr_b v pr_b w
+      | Apply ("LIKE", [v; w]) -> Format.fprintf ppf "(%a) LIKE (%a)" pr_b v pr_b w
+      | Apply (f, args) when SqlFuns.is f -> 
+        let pp_value ppf x = Format.fprintf ppf "%a" pr_b x in
+        let pp_values ppf values = Format.fprintf ppf "%a" (Format.pp_print_list ~pp_sep:pp_comma pp_value) values in
+        Format.fprintf ppf "%a(%a)" pp_constant (SqlFuns.name f) pp_values args
+      | Apply (f, args) -> 
+        let pp_value ppf x = Format.fprintf ppf "%a" pr_b x in
+        let pp_values ppf values = Format.fprintf ppf "%a" (Format.pp_print_list ~pp_sep:pp_comma pp_value) values in
+        Format.fprintf ppf "%a(%a)" pp_constant f pp_values args
+      | Empty q -> Format.fprintf ppf "not exists (%a)" pr_q q
+      | Length q -> Format.fprintf ppf "select count(*) from (%a) as %a" pr_q q pp_constant (fresh_dummy_var ())
+      | RowNumber [] -> Format.fprintf ppf "%a" pp_constant "1"
+      | RowNumber ps -> Format.fprintf ppf "row_number() over (order by %a)" pp_constant (String.concat "," (List.map (string_of_projection quote one_table) ps))
 
 let string_of_base quote one_table b =
   let buf = Buffer.create 0 in
-  string_of_base buf quote one_table b; Buffer.contents buf
+  let ppf = Format.formatter_of_buffer buf in
+  pr_base ppf buf quote one_table b; Format.pp_print_flush ppf (); Buffer.contents buf
 
 let string_of_query ?(range=None) quote q =
+  let pp_constant ppf s = Format.fprintf ppf "%s" s in
```

Comment:
Instead of creating a buffer etc., I would suggest just using `Format.asprintf "%a" pp v`.